Aikido.Zen.Core/Patches/LLMPatcher.cs

```diff
@@ -0,0 +1,228 @@
+
+                if (!TryExtractTokensFromResult(result, out var tokens))
+                {
+                    LogHelper.ErrorLog(Agent.Logger, $"Failed to extract token usage from LLM result for provider: {assembly}, model: {model}");
+                }
+
+                // Record AI statistics
+                Agent.Instance.Context.OnAiCall(assembly, model, tokens.inputTokens, tokens.outputTokens, context.Route);
+
+
+                // record sink statistics
+                Agent.Instance.Context.OnInspectedCall(
+                    operation: $"{__originalMethod.DeclaringType.Namespace}.{__originalMethod.DeclaringType.Name}.{__originalMethod.Name}",
+                    kind: operationKind,
+                    durationInMs: stopWatch.ElapsedMilliseconds,
+                    attackDetected: false,
+                    blocked: false,
+                    withoutContext: context != null
+                );
+            }
+            catch
+            {
+                // Silently handle any errors to avoid affecting the original LLM call
+                LogHelper.ErrorLog(Agent.Logger, "Error tracking LLM call statistics.");
+            }
+        }
+
+        /// <summary>
+        /// Extracts the cloud provider name
+        /// </summary>
+        /// <param name="searchString">The search string to extract the provider from.</param>
+        /// <param name="provider">The extracted provider name.</param>
+        /// <returns>True if the provider was extracted successfully, false otherwise. Not being used at the moment.</returns>
+        internal static bool TryGetCloudProvider(string searchString, out string provider)
+        {
+            provider = "unknown";
+            searchString = searchString.ToLower();
+            // first the cloud providers
+            if (searchString.Contains("azure"))
+            {
+                provider = "azure";
+                return true;
+            }
+            // then the llm companies
+            if (searchString.Contains("anthropic") || searchString.Contains("claude"))
+            {
+                provider = "anthropic";
+                return true;
+            }
+            if (searchString.Contains("google") || searchString.Contains("gemini"))
+            {
+                provider = "gemini";
+                return true;
+            }
+            if (searchString.Contains("mistral"))
+            {
+                provider = "mistral";
+                return true;
+            }
+            // openai last, since their sdk get's used a lot for other llm providers
+            if (searchString.Contains("openai"))
+            {
+                provider = "openai";
+                return true;
+            }
+            return false;
+        }
+
+        /// <summary>
+        /// Converts a dynamic object to a dictionary using reflection
+        /// </summary>
+        internal static Dictionary<string, object> ConvertObjectToDictionary(object obj)
```

Comment:
maybe best to put in a generic helper file? Seems like something we would use in other places as well